File: cpp/LeetCode/No198.cpp
/*
You are a professional robber planning to rob houses along a street. 
Each house has a certain amount of money stashed
, the only constraint stopping you from robbing each of them is 
that adjacent houses have security systems connected 
and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house
, return the maximum amount of money you can rob tonight without alerting the police.
*/
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> mem(nums.size(), -1);
        return dp(nums, 0, mem);
    }
private:
    int dp(vector<int> &nums, int index, vector<int> &mem){
        if(index>=nums.size()) return 0;
        if(mem[index]!=-1) return mem[index];

        int res = max(nums[index] + dp(nums, index+2, mem), dp(nums, index+1, mem));
        mem[index] = res;
        return res;
    }
};

int main(){
    Solution solution;

    vector<int> nums;

    nums = {1,2,3,1};
    cout << solution.rob(nums) << endl;

    nums = {2,7,9,3,1};
    cout << solution.rob(nums) << endl;

    return 0;
}

File: python/LeetCode Quiz/No198.py
'''
You are a professional robber planning to rob houses along a street. 
Each house has a certain amount of money stashed
, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected
 and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house
, return the maximum amount of money you can rob tonight without alerting the police.
'''
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        mem = dict()
        def robTheHouse(i: int) -> int:
            if i==0:
                return nums[i]
            if i==1:
                return max(nums[i-1], nums[i])
            if i not in mem:
                mem[i] = max(robTheHouse(i-2)+nums[i], robTheHouse(i-1))
            return mem[i]
        return robTheHouse(len(nums)-1)

    def rob2(self, nums: List[int]) -> int:
        temp = [nums[0], max(nums[0], nums[1])]
        for i in range(2, len(nums)):
            temp.append(max(temp[i-2]+nums[i], temp[i-1]))
        return temp[-1]

solution = Solution()
nums = [2,7,9,3,1]
print(solution.rob2(nums))